# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/power/WakeLockSentinel.cpp
# Commit: 82e03a404c2f
# Full Hash: 82e03a404c2fe76b1240dce4fa3d0b6bde909a87
# Author: Vincent Hilla <vhilla@mozilla.com>
# Date: 2023-12-05 21:29:44
# Regressor Bug: 1589554
# File Overlap Count: 1
# Description:
#   Bug 1589554 - Part 2: Implement Screen Wake Lock API. r=dom-core,edgar
#   
#   Depends on D189508
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D189509
# ==============================================================================

diff -r b55991835aec -r 82e03a404c2f dom/power/WakeLockSentinel.cpp
--- a/dom/power/WakeLockSentinel.cpp	Tue Dec 05 15:37:05 2023 +0000
+++ b/dom/power/WakeLockSentinel.cpp	Tue Dec 05 15:37:05 2023 +0000
@@ -4,7 +4,14 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include "mozilla/Assertions.h"
+#include "mozilla/dom/Document.h"
+#include "mozilla/dom/Event.h"
+#include "mozilla/dom/EventBinding.h"
+#include "mozilla/dom/Promise.h"
 #include "mozilla/dom/WakeLockSentinelBinding.h"
+#include "mozilla/Hal.h"
+#include "WakeLockJS.h"
 #include "WakeLockSentinel.h"
 
 namespace mozilla::dom {
@@ -14,8 +21,67 @@
   return WakeLockSentinel_Binding::Wrap(cx, this, aGivenProto);
 }
 
-bool WakeLockSentinel::Released() const { return false; }
+bool WakeLockSentinel::Released() const { return mReleased; }
+
+void WakeLockSentinel::NotifyLockReleased() {
+  MOZ_ASSERT(!mReleased);
+  mReleased = true;
+
+  if (mHoldsActualLock) {
+    MOZ_ASSERT(mType == WakeLockType::Screen);
+    NS_DispatchToMainThread(NS_NewRunnableFunction("ReleaseWakeLock", []() {
+      hal::ModifyWakeLock(u"screen"_ns, hal::WAKE_LOCK_REMOVE_ONE,
+                          hal::WAKE_LOCK_NO_CHANGE);
+    }));
+    mHoldsActualLock = false;
+  }
+
+  EventInit init;
+  init.mBubbles = false;
+  init.mCancelable = false;
+  RefPtr<Event> event = Event::Constructor(this, u"release"_ns, init);
+  DispatchTrustedEvent(event);
+}
 
-already_AddRefed<Promise> WakeLockSentinel::ReleaseLock() { return nullptr; }
+void WakeLockSentinel::AcquireActualLock() {
+  MOZ_ASSERT(mType == WakeLockType::Screen);
+  MOZ_ASSERT(!mHoldsActualLock);
+  mHoldsActualLock = true;
+  NS_DispatchToMainThread(NS_NewRunnableFunction("AcquireWakeLock", []() {
+    hal::ModifyWakeLock(u"screen"_ns, hal::WAKE_LOCK_ADD_ONE,
+                        hal::WAKE_LOCK_NO_CHANGE);
+  }));
+}
+
+// https://w3c.github.io/screen-wake-lock/#the-release-method
+already_AddRefed<Promise> WakeLockSentinel::ReleaseLock(ErrorResult& aRv) {
+  // ReleaseWakeLock will remove this from document.[[ActiveLocks]]
+  RefPtr<WakeLockSentinel> kungFuDeathGrip(this);
+
+  if (!mReleased) {
+    nsCOMPtr<nsIGlobalObject> global = GetOwnerGlobal();
+    if (!global) {
+      aRv.Throw(NS_ERROR_NULL_POINTER);
+      return nullptr;
+    }
+    nsCOMPtr<nsPIDOMWindowInner> window = global->GetAsInnerWindow();
+    if (!window) {
+      aRv.Throw(NS_ERROR_NULL_POINTER);
+      return nullptr;
+    }
+    nsCOMPtr<Document> doc = window->GetExtantDoc();
+    if (!doc) {
+      aRv.Throw(NS_ERROR_NULL_POINTER);
+      return nullptr;
+    }
+    ReleaseWakeLock(doc, this, mType);
+  }
+
+  if (RefPtr<Promise> p =
+          Promise::CreateResolvedWithUndefined(GetOwnerGlobal(), aRv)) {
+    return p.forget();
+  }
+  return nullptr;
+}
 
 }  // namespace mozilla::dom